### PR TITLE
apps:healthz implement healthz URL

### LINKED ISSUE #1412: Add a more complete /healthz endpoint to heketi
### Kind of issue

Feature request

### Observed behavior

The /hello endpoint is simple and works well for a quick curl test but says nothing else about the internal state of the server.

### Expected/desired behavior

Provide a /healthz endpoint that follows common conventions and is able to report on more about the internal state of the server.

### PR DESCRIPTION
**Pull request by Madhu-1**
### What does this PR achieve? Why do we need it?

`healthz` checks the following status in heketi.

* check DB is readable.
* check DB is writable.
* check the gluster nodes are reachable.

Signed-off-by: Madhu Rajanna <mrajanna@redhat.com>

### Does this PR fix issues?
Fixes #1412

### Notes for the reviewer
TODO:
* split the commit in to multiple commits
* add test case

### CODE REVIEW

File: apps/glusterfs/health.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
+package glusterfs
```
Comment (phlogistonjohn) [inline apps/glusterfs/health.go]: This file looks to be missing the normal preamble we have in the source files in heketi.

File: apps/glusterfs/health.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	"github.com/lpabon/godbc"
+)
+
+var clusterHealth = "ClusterHealth"
+
+type dbchecker struct {
+	CurrentTime time.Time
+}
+
+func NewHealthz() *api.Healthz {
```
Comment (phlogistonjohn) [inline apps/glusterfs/health.go]: This function seems to be in an odd location between the type definition and that type's functions. Let's move it so it doesn't disrupt the flow when reading the code as much.

File: apps/glusterfs/health.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	return buffer.Bytes(), err
+}
+
+func (d *dbchecker) Unmarshal(buffer []byte) error {
+	dec := gob.NewDecoder(bytes.NewReader(buffer))
+	return dec.Decode(d)
+}
+
+func (*dbchecker) BucketName() string {
+	return BOLTDB_BUCKET_DBATTRIBUTE
```
Comment (phlogistonjohn) [inline apps/glusterfs/health.go]: If this is using the BOLTDB_BUCKET_DBATTRIBUTE bucket why isn't it using the DbAttributeEntry type? I really don't like the idea of multiple types sharing the same bucket. It will just lead to confusion IMO.

File: apps/glusterfs/health.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	return EntryLoad(tx, d, clusterHealth)
+}
+
+func (a *App) Healthz(w http.ResponseWriter, r *http.Request) {
+	info := NewHealthz()
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	//db Read check
+	err := a.db.View(func(tx *bolt.Tx) error {
```
Comment (phlogistonjohn) [inline apps/glusterfs/health.go]: All of the db checking logic should be encapsulated in it's own function call so that the actual rest endpoint function call is simply formatting and reporting on the status of the various checks but not doing all of them directly.

File: apps/glusterfs/health.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	a.db.Update(func(tx *bolt.Tx) error {
+		h := dbchecker{CurrentTime: time.Now()}
+		return h.Write(tx)
+	})
+	if err == nil {
+		info.DbState = api.DbWrite.String()
+	}
+	//cluster status check
+	err = a.db.View(func(tx *bolt.Tx) error {
+		health := NewNodeHealthCache(0, 0, a.db, a.executor)
```
Comment (phlogistonjohn) [inline apps/glusterfs/health.go]: Don't create a new health status cache. We should re-use the one previously set up when the app is set up. If the cache is disabled, simply report an empty but non-failure result.

File: apps/glusterfs/health_cache.go [hunk 6]
```diff
@@ -171,6 +172,10 @@ func (hc *NodeHealthCache) toProbe() ([]*NodeHealthStatus, error) {
 
 func (s *NodeHealthStatus) update(e executors.Executor) {
 	// TODO: add ability to skip check if node was already recently checked
+
+	if !s.Up {
+		return
+	}
```
Comment (phlogistonjohn) [inline apps/glusterfs/health_cache.go]: This doesn't make any sense to me. If the node is up and then goes down later how will ever detect that?

File: apps/glusterfs/health_cache.go [hunk 7]
```diff
@@ -154,14 +154,15 @@ func (hc *NodeHealthCache) toProbe() ([]*NodeHealthStatus, error) {
...
-			}
 			nhs := &NodeHealthStatus{
 				NodeId: nodeId,
 				Host:   node.Info.Hostnames.Manage[0],
 			}
+			// Ignore if the node is not online
+			if !node.isOnline() {
+				continue
+			}
+			nhs.Up = true
```
Comment (phlogistonjohn) [inline apps/glusterfs/health_cache.go]: I don't understand this change. What was wrong with the previous behavior? How does this change improve it? Because this is a subtle thing this is the kind of change I'd prefer to see in a separate commit with it's own commit message explaining the change (assuming it is still needed).

File: pkg/glusterfs/api/types.go [hunk 8]
```diff
@@ -693,3 +694,53 @@ func ValidateIds(v interface{}) error {
...
 }
+
+type NodeHealthStatus struct {
+	NodeId     string    `json:"node"`
+	Host       string    `json:"host"`
+	Up         bool      `json:"online"`
+	LastUpdate time.Time `json:"updated"`
+}
+
+type DbState uint8
```
Comment (phlogistonjohn) [inline pkg/glusterfs/api/types.go]: I don't think this kind of detail is desirable in the api. Just have an enumerated string type with the ok/faulty info needed. That seems where you were going with the Healthz type, so I suggest simply removing this type from the api, and if you still think you want these types, locate them inside the app/glusterfs (for example).
The important part is not to put anything extra in the API.

File: pkg/glusterfs/api/types.go [hunk 9]
```diff
@@ -693,3 +694,53 @@ func ValidateIds(v interface{}) error {
...
+	switch d {
+	case DbRead:
+		state = "read"
+	case DbWrite:
+		state = "write"
+	}
+	return state
+}
+
+type ClusterState uint8
```
Comment (phlogistonjohn) [inline pkg/glusterfs/api/types.go]: See above.

File: apps/glusterfs/health.go (hunk 1 above)
Comment (Madhu-1) [reply apps/glusterfs/health.go]: add license  path

File: apps/glusterfs/health.go (hunk 2 above)
Comment (Madhu-1) [reply apps/glusterfs/health.go]: removed

File: apps/glusterfs/health.go (hunk 3 above)
Comment (Madhu-1) [reply apps/glusterfs/health.go]: create new db attribute for healthz

File: apps/glusterfs/health.go (hunk 4 above)
Comment (Madhu-1) [reply apps/glusterfs/health.go]: done

File: apps/glusterfs/health.go (hunk 5 above)
Comment (Madhu-1) [reply apps/glusterfs/health.go]: sorry I missed that we  are storing  the heathcache in app, updated code

File: apps/glusterfs/health_cache.go (hunk 7 above)
Comment (Madhu-1) [reply apps/glusterfs/health_cache.go]: i have added  it to todo list to make seperate commit,

currently, we  are saving only online node details  in the slice, for the health
implementation, we need  all nodes  status  irrespective of its state to decide whether
the cluster state is healthy or not.

File: apps/glusterfs/health_cache.go (hunk 6 above)
Comment (Madhu-1) [reply apps/glusterfs/health_cache.go]: updated code,  only if the node is marked as up, then only checking the glusterd status or else just  updating the state

File: pkg/glusterfs/api/types.go (hunk 8 above)
Comment (Madhu-1) [reply pkg/glusterfs/api/types.go]: removed

File: apps/glusterfs/health_cache.go (hunk 7 above)
Comment (phlogistonjohn) [reply apps/glusterfs/health_cache.go]: Ok, makes sense. I do wonder if it would be beneficial to keeping two slices, one for up/healthy nodes and one for down/unhealthy nodes. That way when one needs only the list of healthy nodes code only needs to consult one slice w/o needing any additional conditionals.

File: apps/glusterfs/health_cache.go (hunk 6 above)
Comment (phlogistonjohn) [reply apps/glusterfs/health_cache.go]: I'm still confused. In addition this appears to be a conditional that does nothing. Is the current version complete?

File: apps/glusterfs/healthz.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	return buffer.Bytes(), err
+}
+
+func (d *dbchecker) Unmarshal(buffer []byte) error {
+	dec := gob.NewDecoder(bytes.NewReader(buffer))
+	return dec.Decode(d)
+}
+
+func (*dbchecker) BucketName() string {
+	return BOLTDB_BUCKET_DBHEALTHZ
```
Comment (phlogistonjohn) [inline apps/glusterfs/healthz.go]: To be honest I'm really not sure we want a special bucket just for tihs, especially since we're saving a single value. Wouldn't it be simpler and less code to simply set/get an new key under dbattributes?

@raghavendra-talur : I'm curious if you have an opinion about this. The current approach is functional but feels a little like overkill to me. Could I get a 2nd opinion from you?

File: apps/glusterfs/healthz.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+	if len(health.nodes) == 0 {
+		return
+	}
+	totalUnhealthyNodes := 0
+	for _, n := range health.nodes {
+		if !n.Up {
+			totalUnhealthyNodes++
+		}
+		info.NodesStatus = append(info.NodesStatus, (api.NodeHealthStatus)(*n))
```
Comment (phlogistonjohn) [inline apps/glusterfs/healthz.go]: Are the parens around api.NodeHeathStatus needed? It makes it look like a C type cast but it doesn't look like idomatic go to me.

File: apps/glusterfs/healthz.go [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+	if !MonitorGlusterNodes {
+		if err := json.NewEncoder(w).Encode(info); err != nil {
+			panic(err)
+		}
+		return
+	}
+
+	// cluster status check
+	checkClusterState(a.nhealth, &info)
```
Comment (phlogistonjohn) [inline apps/glusterfs/healthz.go]: I think this would read more logically to do the serialization once (at the bottom of the function) and move this line up into the if-staement (with an inverted condition) like so:
```golang
if MonitorGlusterNodes {
    checkClusterState(a.nhealth, &info)
}
// ... serialize response ...
```

File: apps/glusterfs/healthz.go (hunk 11 above)
Comment (Madhu-1) [reply apps/glusterfs/healthz.go]: yes this is needed for type casting

File: apps/glusterfs/healthz.go (hunk 10 above)
Comment (Madhu-1) [reply apps/glusterfs/healthz.go]: @raghavendra-talur let me  know your thoughts on this

**Comment by phlogistonjohn** [comment]
@Madhu-1 do you plan to continue working on this PR?

**Comment by Madhu-1** [comment]
@phlogistonjohn I was waiting for @raghavendra-talur comments on nhttps://github.com/heketi/heketi/pull/1417#discussion_r258905434

**Comment by raghavendra-talur** [comment]
@Madhu-1 I am not 100% clear about "check the gluster nodes are reachable" check.

- What is the need for the check
- Why are we looking at nodes without giving importance to cluster they belong?

**Comment by raghavendra-talur** [comment]
> @Madhu-1 I am not 100% clear about "check the gluster nodes are reachable" check.
> 
> * What is the need for the check

I got the answer to this in the issue, the requirement is "Check that executor has been able to communicate with nodes in some timeframe"
> * Why are we looking at nodes without giving importance to cluster they belong?


I am not sure of the approach where:
1. none of the nodes are reachable == faulty
2. at least one of the nodes is reachable == healthy

Probably a timestamp of last successful reply(not at /healthz endpoint) would be a better candidate.

**Comment by Madhu-1** [comment]
@raghavendra-talur addressed review comments
 below is the response from curl request

```
curl http://127.0.0.01:8080/healthz |python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   574  100   574    0     0  21259      0 --:--:-- --:--:-- --:--:-- 21259
{
    "cluster-state": "healthy",
    "db-state": "write",
    "nodes": [
        {
            "host": "192.168.10.101",
            "node": "252ced46d6cd6effb07a4688e72c829a",
            "online": true,
            "updated": "2019-03-27T15:59:15.063340783+05:30"
        },
        {
            "host": "192.168.10.102",
            "node": "ad4b1df157ac35018c6e9c21cb77ed5b",
            "online": true,
            "updated": "2019-03-27T15:59:15.4671574+05:30"
        },
        {
            "host": "192.168.10.100",
            "node": "b744aca95e4cd3b21ab6a92d497378f9",
            "online": false,
            "updated": "2019-03-27T15:59:18.559924198+05:30"
        },
        {
            "host": "192.168.10.103",
            "node": "d885c1bce818b58af247a7476b952209",
            "online": true,
            "updated": "2019-03-27T15:59:19.0096784+05:30"
        }
    ]
}
```

**Comment by Madhu-1** [comment]
> > @Madhu-1 I am not 100% clear about "check the gluster nodes are reachable" check.
> > 
> > * What is the need for the check
> 
> I got the answer to this in the issue, the requirement is "Check that executor has been able to communicate with nodes in some timeframe"
> 
> > * Why are we looking at nodes without giving importance to cluster they belong?
> 
> I am not sure of the approach where:
> 
> 1. none of the nodes are reachable == faulty

earlier approach was if a number of not reachable nodes are greater than the total nodes, I was marking cluster as faulty, now changed to code to make cluster state as faulty if all nodes are not reachable
> 2. at least one of the nodes is reachable == healthy
> 
> Probably a timestamp of last successful reply(not at /healthz endpoint) would be a better candidate.

**Comment by raghavendra-talur** [comment]
@Madhu-1 @phlogistonjohn I am happy with the PR as it stands right now. I am however, not clear on what I want from /healthz in particular.

Should healthz be local to heketi's status(that means only the db read/write status)?
Or should healthz of heketi represent the status of the storage in whole?

File: apps/glusterfs/healthz.go (hunk 11 above)
Comment (phlogistonjohn) [reply apps/glusterfs/healthz.go]: It's not. I just checked.

**Comment by phlogistonjohn** [comment]
I've looked over much of the code again and my comments and I stand by pretty much all of what I requested previously from my "requested changes". This is also going to need tests.

**Comment by Madhu-1** [comment]
@phlogistonjohn @raghavendra-talur currently I don't have enough bandwidth to work on this one, please feel free to take my patches if they are useful. Thank you :)

**Comment by phlogistonjohn** [comment]
Effort is stale/abandoned. Closing.
